python/cudf/cudf/core/udf/udf_kernel_base.py

```diff
@@ -0,0 +1,193 @@
+    """
+    Base class for kernels computing the result of `.apply`
+    operations on Series or DataFrame objects.
+    """
+
+    def __init__(self, frame, func, args):
+        self.frame = frame
+        self.func = func
+
+        if not all(is_scalar(arg) for arg in args):
+            raise TypeError("only scalar valued args are supported by apply")
+        self.args = args
+        self.nrt = False
+
+        self.frame_type = self._get_frame_type()
+        self.device_func = cuda.jit(device=True)(self.func)
+
+    @property
+    @abstractmethod
+    def kernel_type(self):
+        """
+        API type launching the kernel, used to break
+        degenerecies in the cache.
+        """
+
+    @abstractmethod
+    def _get_frame_type(self):
+        """
+        Numba type of the frame being passed to the kernel.
+        """
+
+    @abstractmethod
+    def _get_kernel_string(self):
+        """
+        Generate executable string of python that defines a
+        kernel we may retrieve from the context and compile
+        with numba.
+        """
+
+    @abstractmethod
+    def _get_kernel_string_exec_context(self):
+        """
+        Get a dict of globals needed to exec the kernel
+        string.
+        """
+
+    def _construct_signature(self, return_type):
+        """
+        Build the signature of numba types that will be used to
+        actually JIT the kernel itself later, accounting for types
+        and offsets. Skips columns with unsupported dtypes.
+        """
+        if not return_type.is_internal:
+            return_type = CPointer(return_type)
+        else:
+            return_type = return_type[::1]
+        # Tuple of arrays, first the output data array, then the mask
+        return_type = Tuple((return_type, boolean[::1]))
+        offsets = []
+        sig = [return_type, int64]
+        for col in _supported_cols_from_frame(self.frame).values():
+            sig.append(_masked_array_type_from_col(col))
+            offsets.append(int64)
+
+        # return_type, size, data, masks, offsets, extra args
+        sig = void(*(sig + offsets + [typeof(arg) for arg in self.args]))
+
+        return sig
+
+    @_performance_tracking
+    def _get_udf_return_type(self):
+        # present a row containing all fields to the UDF and try and compile
+        compile_sig = (self.frame_type, *(typeof(arg) for arg in self.args))
+
+        # Get the return type. The PTX is also returned by compile_udf, but is not
+        # needed here.
+        with _CUDFNumbaConfig():
+            ptx, output_type = cudautils.compile_udf(self.func, compile_sig)
+
+        if not isinstance(output_type, MaskedType):
+            numba_output_type = numpy_support.from_dtype(np.dtype(output_type))
+        else:
+            numba_output_type = output_type
+
+        result = (
+            numba_output_type
+            if not isinstance(numba_output_type, MaskedType)
+            else numba_output_type.value_type
+        )
```

Comment:
This feels too complicated because we are checking for the MaskedType twice. Can we go straight from `output_type`->`result` without the intermediate `numba_output_type` and avoid having to check this twice?